Sockets - Kelly

kdow opened the pull request:
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | Deciding which classes to make, and which methods and variables would be in each class. 
What was a design decision you made that changed over time over the project? | I changed instance variables, which classes had certain methods, and method parameters many times.
What was a concept you gained clarity on, or a learning that you'd like to share? | I think I have a better understanding of instance variables now.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | An example would be testing that check_availability returns a room if one is available. This is nominal because it's a standard use that we expect to happen.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | An edge case example would be testing that a room can be booked with a start date that is the same date as the end date of the previous booking. This is an edge case because it is testing the boundary of the code.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I didn't always write pseudocode first, but I would write tests before code. It was helpful to write pseudocode when I would get stuck on how to write code to get my test(s) to pass.

kdow committed:
Setup Guard and SimpleCov

kdow committed:
Create Room class

kdow committed:
Create Booking class

kdow committed:
Create Room.list_rooms method

kdow committed:
Add request_reservation method to Booking

kdow committed:
Create reservation_dates and find_reservation methods so reservations can be tracked by date

kdow committed:
Add total_cost and nights methods so a total cost can be provided for a reservation

kdow committed:
Create validate_dates method for raising errors on reservation requests with invalid dates

kdow committed:
Extend functionality of check_availability, returns list of available rooms for date range

kdow committed:
Add exception for reservation attempt when no rooms available

kdow committed:
Revise and add availabilty checking tests for booking

kdow committed:
Create Block class

kdow committed:
Add exception for block init containing at least one room that is unavailable for the given date range

kdow committed:
Add refactors.txt

droberts-sea commented on lib/booking.rb:
```diff
@@ -0,0 +1,56 @@
+      date_range = Reservation.reservation_dates(checkin, checkout)
+      room = check_availability(date_range).first
+      num_nights = Reservation.nights(checkin, checkout)
+      reservation = Hotel::Reservation.new(checkin: checkin, checkout: checkout, nights: num_nights, dates: date_range, room: room)
```
Since the `Reservation` already takes the checkin and checkout dates, it feels redundant to also pass in the number of nights and an array of the dates. Making `Reservation` do that work itself helps keep the interface cleaner.

Moreover, what should `Reservation` do if they don't match?

droberts-sea commented on lib/booking.rb:
```diff
@@ -0,0 +1,56 @@
+      avail_rooms = []
+      @rooms.each do |room|
+        if (taken_rooms.include? room.id) == false
+          avail_rooms << room
```
This is me nitpicking, but I have two comments about the conditional on line 35:
- You don't ever need to check equality with `true` or `false` in an `if` statement
- If you have an `if` statement for checking a negative condition, it can be cleaner to write it using `unless`

You also might be able to use the `.reject` enumerable to clarify this code.

Putting all that together, we arrive at:

```ruby
avail_rooms = @rooms.reject do |room|
  taken_rooms.include? room.id
end
```

droberts-sea commented on lib/booking.rb:
```diff
@@ -0,0 +1,56 @@
+      res_by_date = []
+      date = (Date.parse(date)).to_s
+      @reservations.each do |res|
+        if res.dates.any? date
```
Good use of `.any?` here to keep this clean.

droberts-sea commented on lib/reservation.rb:
```diff
@@ -0,0 +1,43 @@
+
+    def total_cost
+      return @nights * room.cost
+    end
```
I like the decision to make this a method rather than storing the cost as an instance variable. That way if the number of nights or the cost per night were to change (say in a future version of the program), the total cost will automatically be correct.

droberts-sea commented on lib/booking.rb:
```diff
@@ -0,0 +1,56 @@
+    def check_availability(date_range)
+      date_range.pop
+      taken_rooms = []
+      date_range.each do |date|
```
This method feels visually cramped. It might be valuable to use empty lines to space it out and separate distinct pieces of logic, similar to the way you would insert a line break between paragraphs in an essay or email. I would probably add line breaks after lines 32 and 38.

droberts-sea commented on lib/booking.rb:
```diff
@@ -0,0 +1,56 @@
+      @reservations.each do |res|
+        if res.dates.any? date
+          res_by_date << res
+        end
```
This code might fit better as an instance method on `Reservation`, something like `Reservation#includes_date?`. Then this code could be simplified to:

```ruby
return @reservations.select { |res| res.includes_date?(date) }
```

I know we hadn't read POODR ch 4 yet when you wrote this code, but this is what Metz is talking about when she says you should "ask for _what_ instead of telling _how_". It's also an example of tight coupling, in that the implementation here relies on the specifics of how `Reservation` works. Tight coupling and law of demeter violations often go hand in hand.

droberts-sea replied:
If you wanted to take that idea even further, you could imagine writing a method `Room#is_available?(checkin, checkout)` that encompasses a lot of the functionality currently contained in `check_availability` and `find_reservation`.

droberts-sea commented on spec/booking_spec.rb:
```diff
@@ -0,0 +1,136 @@
+    it "Raises an error if there are no available rooms for a given date range" do
+      20.times do
+        @booking.request_reservation("April 1, 2020", "April 5, 2020")
+      end
```
Good job getting this edge case.

droberts-sea commented on spec/booking_spec.rb:
```diff
@@ -0,0 +1,136 @@
+  describe "find_reservation" do
+    it "Returns a collection of Reservations for a specific date" do
+      @booking.request_reservation("May 1, 2020", "May 5, 2020")
+      @booking.request_reservation("May 3, 2020", "May 5, 2020")
```
Missing case: does this method omit reservations that don't include the date?

droberts-sea commented on spec/booking_spec.rb:
```diff
@@ -0,0 +1,136 @@
+  describe "further availability checks" do
+    it "Allows a reservation to be made that starts on the same day another reservation for the same room ends" do
+      @booking.request_reservation("April 10, 2020", "April 15, 2020")
+      request = @booking.request_reservation("April 15, 2020", "April 17, 2020")
```
Nice, this section is very thorough.

droberts-sea commented:
# Hotel
## What We're Looking For

<!--
  Comprehension Question Answers:
  Use your best judgment for all of them, check for reasonable answers on nominal/edge case question.
-->

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | you have a lot of work in `Booking` that might fit better elsewhere
Classes are loosely coupled | some - see inline
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | yes
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | yes
**Wave 3** |
Create a block of rooms | N/A
Check if a block has rooms | N/A
Reserve a room from a block | N/A
Test coverage | N/A
**Fundamentals** |
Names variables, classes and modules appropriately | yes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | mostly - see inline
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | yes
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Good job overall! This code is well-written and well-tested, and does a good job of solving the problem. There are a couple of places where your design could be cleaner, but that's part of the point of this project. It is clear to me that your Ruby fundamentals are right where we want them to be. Keep up the hard work!

kdow committed:
Complete design-activity

kdow committed:
Remove nights and dates parameters from Reservation initialize

kdow committed:
Refactor Booking#check_availability

kdow committed:
Add Reservation#includes_date? and refactor Booking#find_reservation

kdow committed:
Add test case for Booking#find_reservation

kdow committed:
Add test for Room#add_reservation

kdow committed:
Add custom error classes

kdow committed:
Add Booking#create_block

kdow committed:
Add functionality for raising errors on reservation attempts of rooms in blocks

Add Room#is_available, Room#add_block, Booking#find_room and refactor code so that a standard reservation cannot be made for a room set aside in a block

kdow committed:
Add functionality so an error is raised if attempting to create a block that contains a room that is already reserved in a block for the given dates.

kdow committed:
Add Block#rooms_available and split Room#is_available into two methods

kdow committed:
Add Booking#reserve_from_block

kdow committed:
Add test for block reservations showing in find_reservation results

kdow committed:
Add error raising for creating blocks of more than 5 rooms
